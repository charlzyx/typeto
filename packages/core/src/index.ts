export { debugInfo, debugNode, debugSymbol, debugType } from "./debug";
export { getNodeExtraInfo, getPropNodeTypeRefName } from "./getNodeExtraInfo";
export { TypeResolver } from "./TypeResolver";
