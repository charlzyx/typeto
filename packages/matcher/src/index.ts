export * from "./AstMatcher";
export * from "./matchers";
