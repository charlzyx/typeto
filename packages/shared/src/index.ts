// export { debugInfo, debugNode, debugSymbol, debugType } from "./debug";
export { getNodeExtraInfo } from "./getNodeExtraInfo";
