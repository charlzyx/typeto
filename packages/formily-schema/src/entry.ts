import {
  ClassDeclaration,
  Node,
  Project,
  TypeAliasDeclaration,
} from "ts-morph";

// 忽略规则的键
const IgnoreRuleKeys = {
  jsDoc: ["ignore", "WIP", "Draft"],
  classDecorators: ["ignore", "WIP", "Draft"],
  leadingComment: ["@ignore", "@WIP", "@Draft"],
};

import { getNodeExtraInfo } from "@typeto/core";

// 判断是否应跳过该节点
const shouldSkip = (node: ClassDeclaration | TypeAliasDeclaration) => {
  const { decorators, jsDocs, leadingComments } = getNodeExtraInfo(node);

  // 检查装饰器中是否有忽略标签
  const skipByDecorator =
    Object.keys(decorators).findIndex((decoratorName) =>
      IgnoreRuleKeys.classDecorators.includes(decoratorName)
    ) > -1;
  if (skipByDecorator) return true;

  // 检查 jsDoc 中是否有忽略标签
  const skipByJsDoc =
    Object.keys(jsDocs).findIndex((tagName) =>
      IgnoreRuleKeys.jsDoc.includes(tagName)
    ) > -1;
  if (skipByJsDoc) return true;

  // 检查前导注释中是否有忽略标签
  const skipByComment =
    leadingComments
      .join("\n")
      .split(/\s+/)
      .findIndex((word) => IgnoreRuleKeys.leadingComment.includes(word)) > -1;
  if (skipByComment) return true;

  return false; // 如果没有任何忽略标签，返回 false
};

/**
 * 获取符合要求的节点
 *
 * class User { namne: string }
 *
 */
// 获取项目中的 d.ts 节点
export const getEntryNodes = (project: Project) => {
  const definitions: ClassDeclaration[] = [];

  project.getSourceFiles().forEach((sourceFile) => {
    sourceFile.getStatements().forEach((statement) => {
      // 处理类声明
      if (Node.isClassDeclaration(statement)) {
        if (shouldSkip(statement)) return; // 跳过被忽略的节点
        definitions.push(statement);
      }
    });
  });

  return {
    // typings,
    definitions,
  };
};
