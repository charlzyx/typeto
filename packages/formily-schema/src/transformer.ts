import { ISchema } from "@formily/json-schema";
import { getNodeExtraInfo, TypeResolver } from "@typeto/core";
import { ClassDeclaration, Type, Node } from "ts-morph";

export const SCHEMA_REF_PREFIX = "#";

type OutputSchema = ISchema & {
  "x-code"?: string;
  $ref?: string;
};

const KeyMap = {
  Component: "x-component",
  ComponentProps: "x-component-props",
  value: "x-value",
  index: "x-index",
  pattern: "x-pattern",
  display: "x-display",
  validator: "x-validator",
  decorator: "x-decorator",
  decoratorProps: "x-decorator-props",
  component: "x-component",
  componentProps: "x-component-props",
  reactions: "x-reactions",
  content: "x-content",
  data: "x-data",
  visible: "x-visible",
  hidden: "x-hidden",
  disabled: "x-disabled",
  editable: "x-editable",
  readOnly: "x-read-only",
  readPretty: "x-read-pretty",
  compileOmitted: "x-compile-omitted",
};

type ResolveContext = {
  extra: Record<string, string | boolean>;
  refs: WeakMap<Type, string>;
  defs: WeakMap<Node, string>;
  self: string;
};

const resolveDefs = (type: Type, ctx: ResolveContext) => {
  const maybe = type.getSymbol()?.getValueDeclaration();
  const has = ctx.defs.has(maybe);
  if (has) {
    return ctx.defs.get(maybe);
  }
};

const convertProps = (extra: ResolveContext["extra"]) => {
  return Object.keys(extra || {}).reduce((map, key) => {
    map[KeyMap[key] || key] = extra[key];
    return map;
  }, {});
};

const consumeExtra = (ctx: ResolveContext) => {
  const extra = convertProps(ctx.extra);
  ctx.extra = {};
  return extra;
};

const primitiveResolve =
  (basetype: ISchema["type"]) => (type: Type, ctx: ResolveContext) => {
    const extra = consumeExtra(ctx);
    return {
      ...extra,
      type: basetype,
      "x-code": type.getText(),
    } satisfies ISchema;
  };

const literalResolve =
  (basetype: ISchema["type"]) => (type: Type, ctx: ResolveContext) => {
    const extra = consumeExtra(ctx);
    return {
      ...extra,
      type: basetype,
      const: type.getLiteralValue(),
      "x-value": type.getLiteralValue(),
      "x-code": type.getText(),
    } satisfies ISchema;
  };

const schemaResolver = new TypeResolver<ISchema, ResolveContext>()
  .before((type, ctx, resolver) => {
    ctx.extra = type.isObject()
      ? ctx.extra
      : {
          "x-decorator": "FormItem",
          "x-component": "Input",
          ...ctx.extra,
        };
    if (ctx.refs.get(type)) {
      return (SCHEMA_REF_PREFIX + ctx.refs.get(type)) as unknown as ISchema;
    } else {
      const maybe = resolveDefs(type, ctx);
      if (maybe && maybe !== ctx.self)
        return (SCHEMA_REF_PREFIX + maybe) as unknown as ISchema;
    }
  })
  .string(primitiveResolve("string"))
  .stringLiteral(literalResolve("string"))
  .boolean(primitiveResolve("boolean"))
  .booleanLiteral(literalResolve("boolean"))
  .number(primitiveResolve("number"))
  .numberLiteral(literalResolve("number"))
  .void(primitiveResolve("void"))
  .undefined(primitiveResolve("null"))
  .null(primitiveResolve("null"))
  .array((type, ctx, resovler) => {
    ctx.refs.set(type, type.getText());
    const extra = consumeExtra(ctx);
    return {
      ...extra,
      type: "array",
      "x-component": "ArrayTable",
      items: {
        type: "void",
        "x-component": "ArrayTable.Item",
        properties: resovler.resolve(
          type.getArrayElementType(),
          ctx
        ) as ISchema["properties"],
      },
    };
  })
  .union((type, ctx, resolver) => {
    ctx.refs.set(type, type.getText());
    const extra = consumeExtra(ctx);
    const subs = type
      .getUnionTypes()
      .map((subType) => resolver.resolve(subType, ctx));
    return {
      ...extra,
      enum: subs.map((item) => ({ label: item.const, value: item.const })),
    };
  })
  .intersection((type, ctx, resolver) => {
    ctx.refs.set(type, type.getText());
    const extra = consumeExtra(ctx);
    return {
      ...extra,
      ...type
        .getIntersectionTypes()
        .map((subType) => resolver.resolve(subType, ctx))
        .reduce((allOf, sub) => {
          return {
            ...allOf,
            ...sub,
          };
        }, {}),
    };
  })
  .object((type, ctx, resolver) => {
    ctx.refs.set(type, type.getText());
    const extra = consumeExtra(ctx);
    return {
      ...extra,
      type: "object",
      properties: type.getProperties().reduce(
        (map, propSymbol) => {
          const propNode =
            propSymbol.getValueDeclaration() ?? propSymbol.getDeclarations()[0];
          const porpType = propNode
            ? propSymbol.getTypeAtLocation(propNode)
            : propSymbol.getDeclaredType();
          if (propNode) {
            const extra = getNodeExtraInfo(propNode);
            ctx.extra = extra.info;
            if (extra.initialValue) {
              ctx.extra["default"] = extra.initialValue;
            }
          }
          map[propSymbol.getName()] = resolver.resolve(porpType, ctx);
          return map;
        },
        {} as ISchema["properties"]
      ),
    };
  });

export const transformDefinitions = (
  definitions: ClassDeclaration[]
): ISchema => {
  const schemas = {} as Record<string, OutputSchema>;
  const globalDefs = new WeakMap<Node, string>();

  for (const def of definitions) {
    globalDefs.set(def, def.getName());
  }

  for (const def of definitions) {
    const extra = getNodeExtraInfo(def);
    const circularRefs = new WeakMap<Type, string>();
    const schema: OutputSchema = {
      ...extra.info,
      type: "object",
      properties: def.getProperties().reduce((map, prop) => {
        const extra = getNodeExtraInfo(prop);
        const ctxExtra = extra.info;
        if (extra.initialValue) {
          ctxExtra["default"] = extra.initialValue;
        }
        map[prop.getName()] = schemaResolver.resolve(prop.getType(), {
          extra: ctxExtra,
          refs: circularRefs,
          defs: globalDefs,
          self: def.getName(),
        });
        return map;
      }, {}),
    };

    schemas[def.getName()] = schema;
  }

  return schemas;
};
