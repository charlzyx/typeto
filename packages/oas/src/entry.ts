import {
  ClassDeclaration,
  Node,
  Project,
  TypeAliasDeclaration,
} from "ts-morph";

// 忽略规则的键
const IgnoreRuleKeys = {
  jsDoc: ["ignore", "WIP", "Draft"],
  classDecorators: ["ignore", "WIP", "Draft"],
  leadingComment: ["@ignore", "@WIP", "@Draft"],
};

import { getNodeExtraInfo } from "@typeto/core";

// 判断是否应跳过该节点
const shouldSkip = (node: ClassDeclaration | TypeAliasDeclaration) => {
  const { decorators, jsDocs, leadingComments } = getNodeExtraInfo(node);

  // 检查装饰器中是否有忽略标签
  const skipByDecorator =
    Object.keys(decorators).findIndex((decoratorName) =>
      IgnoreRuleKeys.classDecorators.includes(decoratorName)
    ) > -1;
  if (skipByDecorator) return true;

  // 检查 jsDoc 中是否有忽略标签
  const skipByJsDoc =
    Object.keys(jsDocs).findIndex((tagName) =>
      IgnoreRuleKeys.jsDoc.includes(tagName)
    ) > -1;
  if (skipByJsDoc) return true;

  // 检查前导注释中是否有忽略标签
  const skipByComment =
    leadingComments
      .join("\n")
      .split(/\s+/)
      .findIndex((word) => IgnoreRuleKeys.leadingComment.includes(word)) > -1;
  if (skipByComment) return true;

  return false; // 如果没有任何忽略标签，返回 false
};

/**
 * 获取符合要求的节点
 *
 * class User { namne: string }
 *
 * type getUser = {
 *    url: "/get/user"
 * }
 */
// 获取项目中的 d.ts 节点
export const getEntryNodes = (project: Project) => {
  // const typings: TypeAliasDeclaration[] = [];
  const definitions: ClassDeclaration[] = [];
  const operations: TypeAliasDeclaration[] = [];
  const unique = {} as Record<string, TypeAliasDeclaration>;

  project.getSourceFiles().forEach((sourceFile) => {
    sourceFile.getStatements().forEach((statement) => {
      // 处理类声明
      if (Node.isClassDeclaration(statement)) {
        if (shouldSkip(statement)) return; // 跳过被忽略的节点
        definitions.push(statement);
      }
      // 处理类型别名声明
      if (Node.isTypeAliasDeclaration(statement)) {
        const typ = statement.getTypeNode();
        if (shouldSkip(statement)) return; // 跳过被忽略的节点

        // 如果是带有泛型的说明是一些辅助类型定义, 虽然暂时没有什么用, 但是先保起来
        // const typeParams = statement.getTypeParameters();
        // 仅处理有泛型参数的类型别名
        // if (typeParams.length > 0) {
        //   typings.push(statement);
        // }

        // 否则就看是不是 ApiOperation 定义
        // 条件是必须是 TypeLiteral 字面量定义
        if (Node.isTypeLiteral(typ)) {
          const operation = statement;
          // 必须有 url 字段
          const hasUrl = operation.getType().getProperty("url");
          if (hasUrl) {
            const name = operation.getName();
            if (unique[name]) {
              throw new Error(
                "Api 操作定义出现了重复" +
                  name +
                  "重复定义所在文件: " +
                  unique[name].getSourceFile().getFilePath() +
                  ":" +
                  unique[name].getStartLineNumber() // 使用 getLine() 获取行号
              );
            }
            operations.push(operation);
            unique[name] = operation; // 记录唯一的操作
          }
        }
      }
    });
  });

  return {
    // typings,
    definitions,
    operations,
  };
};
