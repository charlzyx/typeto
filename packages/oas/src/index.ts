import { Project } from "ts-morph";
// import { transformDefinitions, transformOperations } from "./transformer";
import { transformDefinitions, transformOperations } from "./transformer2";
import { getEntryNodes } from "./entry";

const clean = (x: any) =>
  JSON.parse(
    JSON.stringify(x, (k: string, v) => {
      if (k === "_code") return undefined;
      return v;
    })
  );

export const transform = (project: Project) => {
  const { definitions, operations } = getEntryNodes(project);
  const { defSchema, defNameMap } = transformDefinitions(definitions);
  const opSchema = transformOperations(operations, defNameMap);
  return { operations: clean(opSchema), definitions: clean(defSchema) };
};
