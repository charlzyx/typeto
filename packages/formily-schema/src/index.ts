import { Project } from "ts-morph";
import { transformDefinitions } from "./transformer";
import { getEntryNodes } from "./entry";
import { ISchema } from "@formily/json-schema";

const clean = (schema: ISchema) => {
  const replacer = (k: string, v: any) => {
    if (k === "x-code") return undefined;
    return v;
  };

  return JSON.parse(JSON.stringify(schema, replacer));
};

export const transform = (project: Project) => {
  const { definitions } = getEntryNodes(project);
  const schema = transformDefinitions(definitions);
  const schemaDefs = clean(schema);
  const prefix = 'import { ISchema } from "@formily/json-schema";';
  const reforamt = Object.keys(schemaDefs).map((def) => {
    return `const ${def}: ISchema = ${JSON.stringify(
      schemaDefs[def],
      null,
      2
    )};\n`.replace(/"#\w+"/g, (m) => {
      return m.replace(/"/g, "").replace("#", "");
    });
  });
  return [prefix, ...reforamt].join("\n");
};
