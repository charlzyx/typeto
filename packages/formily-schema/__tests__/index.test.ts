import { describe, expect, it } from "bun:test";
import { Project } from "ts-morph";
import { transform } from "../src/index";
import { getEntryNodes } from "../src/entry";

describe("formily schema tests", () => {
  const project = new Project({});

  it("应该正确转换用户信息模式 / should transform user information schema correctly", () => {
    const sourceFile = project.createSourceFile(
      "case1.ts",
      `
  class Form {
    @title(用户信息)
    user: User;
  }

  class User {
    @title(用户名)
    name: string;
    /** 地址信息 */
    address: Address;
    /** 地址列表 */
    addressList: Address[]
  }

  class Address {
    /** @title 城市 */
    city: "北京" | "上海";
    @title(街道)
    street: string;
  }
        `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();

    project.removeSourceFile(sourceFile);
  });

  it("应该跳过被忽略的节点 / should skip ignored nodes", () => {
    const sourceFile = project.createSourceFile(
      "case2.ts",
      `
  class Form {
    @ignore
    ignoredClass: IgnoredClass;

    visibleClass: VisibleClass;
  }

  @ignore
  class IgnoredClass {
    name: string;
  }

  class VisibleClass {
    age: number;
  }
        `
    );

    const { definitions } = getEntryNodes(project);
    expect(
      definitions.find((x) => x.getName() == "IgnoredClass")
    ).toBeUndefined(); // 只应返回 VisibleClass

    project.removeSourceFile(sourceFile); // 删除虚拟文件
  });

  it("应该处理循环引用 / should handle circular references", () => {
    const sourceFile = project.createSourceFile(
      "case3.ts",
      `
class Form {
  a: A;
}

class A {
  b: B;
}

class B {
  a: A;
}
      `
    );

    const result = transform(project);

    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });

  it("应该正确转换基本类型 / should transform basic types correctly", () => {
    const sourceFile = project.createSourceFile(
      "case4.ts",
      `
  class Form {
    basicTypes: BasicTypes;
  }

  class BasicTypes {
    str: string;
    num: number;
    bool: boolean;
  }
        `
    );

    const result = transform(project);

    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });
});
