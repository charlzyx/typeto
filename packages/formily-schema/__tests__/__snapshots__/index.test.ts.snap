// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`formily schema tests 应该正确转换用户信息模式 / should transform user information schema correctly 1`] = `
"import { ISchema } from "@formily/json-schema";
const Form: ISchema = {
  "type": "object",
  "properties": {
    "user": User
  }
};

const User: ISchema = {
  "type": "object",
  "properties": {
    "name": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "title": "用户名",
      "type": "string"
    },
    "address": Address,
    "addressList": {
      "description": "地址列表",
      "type": "array",
      "x-component": "ArrayTable",
      "items": {
        "type": "void",
        "x-component": "ArrayTable.Item",
        "properties": "#Address"
      }
    }
  }
};

const Address: ISchema = {
  "type": "object",
  "properties": {
    "city": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "title": "城市",
      "enum": [
        {
          "label": "北京",
          "value": "北京"
        },
        {
          "label": "上海",
          "value": "上海"
        }
      ]
    },
    "street": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "title": "街道",
      "type": "string"
    }
  }
};
"
`;

exports[`formily schema tests 应该处理循环引用 / should handle circular references 1`] = `
"import { ISchema } from "@formily/json-schema";
const Form: ISchema = {
  "type": "object",
  "properties": {
    "a": A
  }
};

const A: ISchema = {
  "type": "object",
  "properties": {
    "b": B
  }
};

const B: ISchema = {
  "type": "object",
  "properties": {
    "a": A
  }
};
"
`;

exports[`formily schema tests 应该正确转换基本类型 / should transform basic types correctly 1`] = `
"import { ISchema } from "@formily/json-schema";
const Form: ISchema = {
  "type": "object",
  "properties": {
    "basicTypes": BasicTypes
  }
};

const BasicTypes: ISchema = {
  "type": "object",
  "properties": {
    "str": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "type": "string"
    },
    "num": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "type": "number"
    },
    "bool": {
      "x-decorator": "FormItem",
      "x-component": "Input",
      "type": "boolean"
    }
  }
};
"
`;
